Separate WGs and Projects lifecycles (#95)
ossf/tac: process/project-lifecycle.md

# Project Life Cycle

> Projects are either directly under the governance of the Technical Advisory Committee (TAC) or a specific Working Group (WG). The progression of a project lies accordingly within the TAC or the specific WG the project reports to (i.e., "the parent WG").

Reviewer: The TAC should be described as the responsible party -- a WG can recommend a project move to specific maturity levels, but the TAC must approve it. 

This provides consistency to all external parties who will rely on the OpenSSF *as a whole* to signal in a consistent manner, and it aligns with the Charter's definition of the TAC's role.

Proposed wording:

```
Projects are either directly under the governance of the Technical Advisory Committee (TAC) or a specific Working Group (WG). The TAC is responsible for reviewing and approving changes to all Project's Life Cycles. Where a Project reports into a specific WG, that WG can support the Project's progression and provide recommendations to the TAC.

Reviewer 2: nit: s/approving changes/approving significant changes
I don't think we would have time to look at all changes.

Author: Done.

Projects follow the sandbox, incubating, graduated and archival lifecycle stagesdefined below.

> The lifecycle progression, movement from one stage to another, allows them to participate at the level that is most appropriate for them given where they are at.

Reviewer: ```suggestion
The lifecycle progression, movement from one stage to another, allows them to participate at the level that is most appropriate for them given where they are at, and signals to our community and to those outside of it what level of maturity and continuity they can expect from a Project.
```

Reviewer 2: +1

Author: Done.

      - [Archived](#archived)

<!-- /TOC -->

## Stages - Definitions & Expectations

Projects have a maturity level of sandbox, incubating, or graduated. Archived is for Projects no longer in active development. The maturity level is a signal by OpenSSF as to what sorts of enterprises should be adopting different projects. Projects increase their maturity by demonstrating their sustainability to OpenSSF’s Technical Advisory Committee(TAC): that they have adoption, a healthy rate of changes, committers from multiple organizations, have adopted the OpenSSF Code of Conduct, and have achieved and maintained the Core Infrastructure Initiative Best Practices Badge.

![Project Stages](/process/project-stages.png)

### Sandbox

The OpenSSF Sandbox is the entry point for early stage Projects and has four goals:

* Encourage public visibility of experiments or other early work that can add value to the OpenSSF mission and build the ingredients of a successful Incubation level Project
* Facilitate alignment with existing Working Groups or Projects if (and only if) this is desired
* Nurture Projects (e.g. via OpenSSF Service Desk requests)

> * Remove possible legal and governance obstacles to adoption and contribution by ensuring all projects adhere to OpenSSF legal, code of conduct and IP Policy requirements.

Reviewer 2: +1

#### Gives - Review and Reporting
*** Need Text ***

#### Gets - Benefits
*** Need Text ***

#### Sandbox Entry Requirements

> Projects being submitted to the OpenSSF at the sandbox level are intended to be the entry point for early stage projects and are not required to undergo due diligence.

Reviewer 2: I feel this is too strong of a requirement for an early stage project (I do agree that this is required for mature, company owned projects). We should keep as-is.

> Projects being submitted to the OpenSSF at the sandbox level are intended to be the entry point for early stage projects and are not required to undergo due diligence.

Author: I've left it as-is for now.

> Projects being submitted to the OpenSSF at the sandbox level are intended to be the entry point for early stage projects and are not required to undergo due diligence.

Reviewer 3: I think the Sandbox entry requirements and process (eg how due diligence is defined at this stage) is important for putting whatever the OpenSSF's vision for a "project landscape" is into practice. If the OpenSSF intends to be more opinionated, and donation/new projects is done _in service_ of supporting upstream security changes, then I think a very lenient Sandbox entry would be misleading to maintainers (whose projects may be stuck in sandbox indefinitely). If the intention is to cultivate many new technologies in the space (a la CNCF), then a lenient Sandbox entry is more appropriate. In general, I have mostly heard the former being OpenSSF's POV on project donation, but I don't think I have yet seen this formalized through the TAC or GB.

> Projects being submitted to the OpenSSF at the sandbox level are intended to be the entry point for early stage projects and are not required to undergo due diligence.

Reviewer: Defining the scope/vision of the OpenSSF, so that we might evaluate projects against that, is a good discussion for the TAC to have. However, my initial point that IP Transfer is required for the sandbox stage stands. 

> I feel this is too strong of a requirement for an early stage project

@inferno-chromium , this isn't about feelings -- it is a legal matter. The LF cannot accept a pre-existing project without an IP Transfer agreement in place between the contributing company and the LF. Whomever holds the copyright on the project code, or trademark on its name, regardless of its maturity, must come to an agreement with the LF... or it is not an LF-managed project, and therefore would not be part of the OpenSSF and should not be using the OpenSSF name or brand.

> Projects being submitted to the OpenSSF at the sandbox level are intended to be the entry point for early stage projects and are not required to undergo due diligence.

Reviewer 2: @AevaOnline - Be respectful in your replies and please don't try to enforce your opinion over others. Lets other TAC members chime in as well. A definition of early stage project is important here and that will decide what needs to be put in place.

Sandbox Projects should be early-stage projects that the OpenSSF TAC or parent Working Group believes warrant experimentation.

* New Projects that are designed to extend one or more OpenSSF Projects with functionality or interoperability libraries.
* Independent projects that fit the OpenSSF mission and provide potential for a novel approach to existing functional areas (or are an attempt to meet an unfulfilled need)
* Projects commissioned or sanctioned by the OpenSSF, including initial code for OpenSSF WG collaborations, and "experimental" projects
* Any project that realistically intends to join OpenSSF Incubation in future and wishes to lay the foundations for that

#### Project Graduation Process: Sandbox to Incubating

> Incubating Projects are required to undergo due diligence as a part of the process to move from Sandbox to Incubation. Due Diligence is driven by a TAC or parent WG sponsor, with two weeks for public comment before a vote is called.

Reviewer: ```suggestion
Incubating Projects are required to undergo due diligence as a part of the process to move from Sandbox to Incubation. Due Diligence is driven by a TAC or parent WG sponsor. Once the diligence is complete and the proposal made, two weeks are allocated for public comment before a TAC vote is called.
```

Reviewer 2: +1

#### Submission Forms for Sandbox

1. [Proposal](TBD)
2. [Legal](TBD)

#### Submission Forms for Sandbox to Incubating

1. [Proposal](TBD)
2. [Legal](TBD)

### Incubating

![Incubating](/process/incubation-process.png)

#### Gives - Review and Reporting
*** Need Text ***

#### Gets - Benefits
*** Need Text ***

#### Incubating Entry Requirements

To be accepted to incubating stage, a project must meet the sandbox stage requirements plus:

* Document that it is being used successfully in production by at least three independent end users which, in the TAC’s judgement, are of adequate quality and scope.

> * Have a healthy number of committers. A committer is defined as someone with the commit bit; i.e., someone who can accept contributions to some or all of the project.

Reviewer: ```suggestion
* Have a healthy number of maintainers. A maintainer is defined as someone with the commit bit; i.e., someone who can accept and merge contributions to some or all of the project.
```

Author: Done.

> * Demonstrate a substantial ongoing flow of commits and merged contributions.

Reviewer: ```suggestion
* Demonstrate a substantial ongoing flow of commits and merged contributions.
* Have maintainers (those with merge rights) from at least two OpenSSF Members.
```

* Since these metrics can vary significantly depending on the type, scope and size of a project, the TAC has final judgement over the level of activity that is adequate to meet these criteria
* A clear versioning scheme.
* Specifications must have at least one public reference implementation.

> #### Project Graduation Process: Incubating to Graduation

Author: The way the document is currently structured is:
State n
Graduation process from stage n to stage n+1
State n+1
....

I take it that you would prefer:
State n
State n+1
Graduation process from stage n to stage n+1
....

I think we need to decide on the structure and be consistent throughout. I can go either way.

Projects that wish to move from Incubating to Graduation should open a PR confirming the following criteria:

> * Have committers from at least two organizations.

Reviewer: ```suggestion
* Have committers from at least three independent (where one is not a subsidiary of the other) organizations.
```

Reviewer: Yes, I'm proposing the bar be set at 3, which is where the CNCF originally set the bar. Happy to discuss this as it is a significant change.

Reviewer 2: 3 feels too strong. Maybe 2 with a few other open source contributors. Happy to discuss this in tac meeting, but seems fine to leave 2 for now.

* Have achieved and maintained a [Core Infrastructure Initiative Best Practices Badge](https://bestpractices.coreinfrastructure.org/).

> * Have completed an independent and third party security audit with results published of similar scope and quality as [this example](https://github.com/envoyproxy/envoy#security-audit) which includes all critical vulnerabilities and all critical vulnerabilities need to be addressed before graduation.

Reviewer: ```suggestion
* Have completed an independent and third party security audit with results published of similar scope and quality as [this example](https://github.com/envoyproxy/envoy#security-audit) which includes all critical vulnerabilities and all critical vulnerabilities need to be addressed before graduation. The TAC, or another WG, may provide assistance in acquiring such an audit.
```

> * Explicitly define a project governance and committer process. This preferably is laid out in a GOVERNANCE.md file and references an OWNERS.md file showing the current and emeritus committers.

Reviewer: ```suggestion
* Explicitly define project governance structure and process to be granted maintainership within the project team. This preferably is laid out in a GOVERNANCE.md file and references an OWNERS.md file showing the current and emeritus maintainers.
```

Reviewer 2: +1

Author: Done.

# Project Life Cycle

> Projects are either directly under the governance of the Technical Advisory Committee (TAC) or a specific Working Group (WG). The TAC is responsible for reviewing and approving all significant changes to a Project's life cycle. Where a Project reports into a specific WG, that WG can support the Project's progression and provide recommendations to the TAC.

Reviewer 3: The second sentence here could be read as contradicting the first. Is the intended workflow that all net new projects receive TAC approval, and then after that approval, some may report up to WGs instead of the TAC if there is a relevant WG?

Author: No, the intent is that whether projects report to the TAC or a WG their lifecycle is under the control of the TAC (I think this is however not a settled issue). I'm clarifying the text though. Thanks.

![Project Stages](/process/project-stages.png)

### Sandbox

The OpenSSF Sandbox is the entry point for early stage Projects and has four goals:

* Encourage public visibility of experiments or other early work that can add value to the OpenSSF mission and build the ingredients of a successful Incubation level Project
* Facilitate alignment with existing Working Groups or Projects if (and only if) this is desired
* Nurture Projects (e.g. via OpenSSF Service Desk requests)
* Remove possible legal and governance obstacles to adoption and contribution by ensuring all projects adhere to OpenSSF legal, code of conduct, and IP Policy requirements.

#### Gives - Review and Reporting
*** Need Text ***

#### Gets - Benefits

> *** Need Text ***

Reviewer 3: I think OpenSSF might do well to borrow the ASF approach for [branding of podling vs incubators](https://incubator.apache.org/guides/branding.html). Because of the momentum behind open source security at the moment, I think reserving some marketing benefits until projects reach an incubation or graduation stage would help keep the focus on developing the technology of the Sandbox projects and encourages projects to progress.

Reviewer 4: Agreed.

Author: I'm leaving this to be addressed by PR#100.

* Projects commissioned or sanctioned by the OpenSSF, including initial code for OpenSSF WG collaborations, and "experimental" projects
* Any project that realistically intends to join OpenSSF Incubation in future and wishes to lay the foundations for that

#### Project Graduation Process: Sandbox to Incubating

Incubating Projects are required to undergo due diligence as a part of the process to move from Sandbox to Incubation. Due Diligence is driven by a TAC or parent WG sponsor. Once the diligence is complete and the proposal made, two weeks are allocated for public comment before a TAC vote is called.

#### Submission Forms for Sandbox

1. [Proposal](TBD)
2. [Legal](TBD)

#### Submission Forms for Sandbox to Incubating

1. [Proposal](TBD)
2. [Legal](TBD)

### Incubating

![Incubating](/process/incubation-process.png)

#### Gives - Review and Reporting
*** Need Text ***

#### Gets - Benefits
*** Need Text ***

#### Incubating Entry Requirements

To be accepted to incubating stage, a project must meet the sandbox stage requirements plus:

> * Document that it is being used successfully in production by at least three independent end users which, in the TAC’s judgement, are of adequate quality and scope.

Reviewer 4: is the intent to only permit end users as a qualifier?  Depending on the nature of a given project in fixing upstream security issues, i can't imagine we're terribly concerned about end users, rather we'd like to focus on adopters which may include other open source security projects who inherently are not end users. (This may be reuse from other foundations where End Users are official things within the construct of the foundation).

Author: Makes sense to me. Changing to "adopters. Thanks.

> * Have a healthy number of maintainers. A maintainer is defined as someone with the commit bit; i.e., someone who can accept and merge contributions to some or all of the project.

Reviewer 4: What is healthy in regards to this?  with this concept of "healthy number" also be reused as a criteria for the dashboard or criticality score? If not, why not?

Author: I think that's a valid point that should be addressed independently. I will open an issue. Thanks.

Author: See issue #101

* Demonstrate a substantial ongoing flow of commits and merged contributions.
* Since these metrics can vary significantly depending on the type, scope and size of a project, the TAC has final judgement over the level of activity that is adequate to meet these criteria
* A clear versioning scheme.
* Specifications must have at least one public reference implementation.

#### Project Graduation Process: Incubating to Graduation

Projects that wish to move from Incubating to Graduation should open a PR confirming the following criteria:
* Have maintainers from at least two organizations.

> * Have achieved and maintained a [Core Infrastructure Initiative Best Practices Badge](https://bestpractices.coreinfrastructure.org/).

Author: Indeed. Thanks!
